presto-main/src/test/java/com/facebook/presto/sql/planner/PlanAssert.java

```diff
@@ -0,0 +1,415 @@
+            if (anyTree) {
+                int sourcesCount = node.getSources().size();
+                if (sourcesCount > 1) {
+                    states.add(new PlanMatching(nCopies(sourcesCount, this), symbolAliases));
+                }
+                else {
+                    states.add(new PlanMatching(ImmutableList.of(this), symbolAliases));
+                }
+            }
+            if (node.getSources().size() == sources.size() && matchers.stream().allMatch(it -> it.matches(node, session, metadata, symbolAliases))) {
+                states.add(new PlanMatching(sources, symbolAliases));
+            }
+            return states.build();
+        }
+
+        public PlanMatcher withSymbol(String pattern, String alias)
+        {
+            return with(new SymbolMatcher(pattern, alias));
+        }
+
+        public PlanMatcher with(Matcher matcher)
+        {
+            matchers.add(matcher);
+            return this;
+        }
+
+        public PlanMatcher matchToAnyNodeTree()
+        {
+            anyTree = true;
+            return this;
+        }
+
+        private boolean isTerminated()
+        {
+            return sources.isEmpty();
+        }
+    }
+
+    private static class SymbolMatcher
+            implements Matcher
+    {
+        private final Pattern pattern;
+        private final String alias;
+
+        public SymbolMatcher(String pattern, String alias)
+        {
+            this.pattern = Pattern.compile(pattern);
+            this.alias = alias;
+        }
+
+        @Override
+        public boolean matches(PlanNode node, Session session, Metadata metadata, SymbolAliases symbolAliases)
+        {
+            Symbol symbol = null;
+            for (Symbol outputSymbol : node.getOutputSymbols()) {
+                if (pattern.matcher(outputSymbol.getName()).find()) {
+                    checkState(symbol == null, "%s symbol was found multiple times in %s", pattern, node.getOutputSymbols());
+                    symbol = outputSymbol;
+                }
+            }
+            if (symbol != null) {
+                symbolAliases.put(alias, symbol);
+                return true;
+            }
+            return false;
+        }
+    }
+
+    public interface Matcher
+    {
+        boolean matches(PlanNode node, Session session, Metadata metadata, SymbolAliases symbolAliases);
+    }
+
+    private static class PlanMatching
+    {
+        private final List<PlanMatcher> planMatchers;
+        private final SymbolAliases symbolAliases;
+
+        private PlanMatching(List<PlanMatcher> planMatchers, SymbolAliases symbolAliases)
+        {
+            requireNonNull(symbolAliases, "symbolAliases is null");
+            requireNonNull(planMatchers, "matchers is null");
+            this.symbolAliases = new SymbolAliases(symbolAliases);
+            this.planMatchers = ImmutableList.copyOf(planMatchers);
+        }
+
+        public boolean isTerminated()
```

Comment:
`PlanMatching` (I will rename it) is conceptually similar to state in the state machine. So when it is terminated it means that we reached the last state and we have partial match. To have plan match, all the plan branches have to reach this terminated state.